### PR TITLE
Refactor Automatic Serialization API

### PR DESCRIPTION
**Pull request by rdspring1**
This PR fixes issues encountered when deserializing an incompatible workspace, which can occur when pulling the latest changes or switching NvFuser branches.

1. Add `enable_automatic_serialization` and `disable_automatic_serialization` functions to `nvfuser/__init__.py`
2. Move automatic deserialization from `FusionCache` constructor to `FusionCache::get()`.
3. Expand try-catch block to handle `FusionCache::deserialize`. If an exception is detected in common workspace, reset `Fusion Cache`.
4. Detect exceptions in thread pool workers, and then throw an exception after the thread pool finishes its work.

### CODE REVIEW

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

File: csrc/kernel_cache.cpp [hunk 1]
```diff
@@ -1218,11 +1220,22 @@ void FusionKernelRuntime::compileFusionParallel(KernelArgumentHolder args) {
...
-        compileKernel(group_runtime_inputs, group_to_run);
+        try {
+          c10::cuda::CUDAGuard dg(args.getDeviceIndex());
+          c10::Device device(c10::DeviceType::CUDA, args.getDeviceIndex());
+          compileKernel(group_runtime_inputs, group_to_run);
+        } catch (const std::exception& e) {
+          // Set flag inside lambda so we can throw an exception after thread
+          // pool completes its work.
+          std::cout << e.what() << std::endl;
+          detect_exception_in_thread_pool = true;
```
Comment (jjsjann123) [inline csrc/kernel_cache.cpp:1237]: nitpick: thread safety. use atomic or mutex.

File: csrc/python_frontend/fusion_cache.cpp [hunk 2]
```diff
@@ -89,24 +90,21 @@ BinaryBuffer openFusionCache(std::string filename) {
...
-    bool strict) {
+const serde::FusionCache* verifyFusionCache(const BinaryBuffer& buffer) {
   FUSER_PERF_SCOPE("Flatbuffers::verifyFusionCache");
   auto fusion_cache_buffer = serde::GetFusionCache(buffer.data());
 
   // Check flatbuffer integrity
   flatbuffers::Verifier v(buffer.data(), buffer.size());
   if (!fusion_cache_buffer->Verify(v)) {
-    NVF_CHECK(!strict, "Failed to verify the integrity of FusionCache buffer.");
+    NVF_CHECK(false, "Failed to verify the integrity of FusionCache buffer.");
```
Comment (jjsjann123) [inline csrc/python_frontend/fusion_cache.cpp:100]: nitpick: make the if condition an NVF_CHECK instead.

File: csrc/python_frontend/fusion_cache.cpp [hunk 3]
```diff
@@ -89,24 +90,21 @@ BinaryBuffer openFusionCache(std::string filename) {
...
     return nullptr;
   }
 
   // Check schema version
   if (!serde::FusionCacheBufferHasIdentifier(buffer.data())) {
     NVF_CHECK(
-        !strict,
-        "Failed to verify the schema version of the FusionCache buffer");
+        false, "Failed to verify the schema version of the FusionCache buffer");
```
Comment (jjsjann123) [inline csrc/python_frontend/fusion_cache.cpp:107]: ditto

File: csrc/python_frontend/fusion_cache.cpp [hunk 4]
```diff
@@ -115,7 +113,7 @@ const serde::FusionCache* verifyFusionCache(
   if (device_prop->major != fusion_cache_buffer->device_major() ||
       device_prop->minor != fusion_cache_buffer->device_minor()) {
     NVF_CHECK(
-        !strict,
+        false,
```
Comment (jjsjann123) [inline csrc/python_frontend/fusion_cache.cpp:116]: ditto

File: csrc/python_frontend/fusion_cache.cpp [hunk 5]
```diff
@@ -134,7 +132,7 @@ const serde::FusionCache* verifyFusionCache(
   if (cuda_major != fusion_cache_buffer->cuda_major() ||
       cuda_minor != fusion_cache_buffer->cuda_minor()) {
     NVF_CHECK(
-        !strict,
+        false,
```
Comment (jjsjann123) [inline csrc/python_frontend/fusion_cache.cpp:135]: ditto

File: csrc/python_frontend/fusion_cache.cpp [hunk 6]
```diff
@@ -710,10 +711,17 @@ void FusionCache::deserialize(
...
-        fusion_schedule->auto_gen_schedules->deserialize(
-            fb_fec_node, (int64_t)trie_node->fusion_id);
+        try {
+          fusion_schedule->auto_gen_schedules->deserialize(
+              fb_fec_node, (int64_t)trie_node->fusion_id);
+        } catch (const std::exception& e) {
+          // Set flag inside lambda so we can throw an exception after thread
+          // pool completes its work.
+          std::cout << e.what() << std::endl;
+          detect_exception_in_thread_pool = true;
```
Comment (jjsjann123) [inline csrc/python_frontend/fusion_cache.cpp:723]: thread safety warning~

File: nvfuser/__init__.py [hunk 7]
```diff
@@ -33,6 +33,20 @@
...
 
+# Register automatic serialization of Nvfuser cache hierarchy and cuda kernels.
+def enable_automatic_serialization():
+    import atexit
+
+    atexit.register(_C.serialize)
+
+
+# Unregister automatic serialization of Nvfuser cache hierarchy and cuda kernels.
+def disable_automatic_serialization():
```
Comment (jjsjann123) [inline nvfuser/__init__.py:44]: This thing isn't used.

Comment (rdspring1) [reply nvfuser/__init__.py:44]: This is to give users the option to disable automatic serialization in Thunder.

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Review by jjsjann123 — APPROVED** [review]
LGTM.

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by jjsjann123** [comment]
hmm. what's with the segmentation error here?
```
 test_squeeze (__main__.TestNvFuserFrontend) ... [xpl-dvt-52:477  :0:477] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x70)
==== backtrace (tid:    477) ====
 0 0x0000000000042520 __sigaction()  ???:0
 1 0x000000000030277c nvfuser::FusionExecutor::serialize()  ???:0
 2 0x00000000003096a2 nvfuser::FusionExecutor::serialize()  ???:0
 3 0x000000000044eec2 nvfuser::FusionKernelRuntime::serialize()  ???:0
 4 0x00000000004578cc nvfuser::FusionExecutorCache::serialize()  ???:0
 5 0x000000000073a9d6 nvfuser::python_frontend::FusionCache::serialize()  ???:0
 6 0x00000000000f83ec pybind11::cpp_function::initialize<nvfuser::python_frontend::initNvFuserPythonBindings(_object*)::{lambda(nvfuser::python_frontend::FusionCache&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)#1}, void, nvfuser::python_frontend::FusionCache&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, pybind11::name, pybind11::is_method, pybind11::sibling, pybind11::arg>(nvfuser::python_frontend::initNvFuserPythonBindings(_object*)::{lambda(nvfuser::python_frontend::FusionCache&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)#1}&&, void (*)(nvfuser::python_frontend::FusionCache&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&, pybind11::arg const&)::{lambda(pybind11::detail::function_call&)#3}::_FUN()  python_bindings.cpp:0
```

That looks a bit concerning....

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
I'm not sure about the segmentation error on the CI. It doesn't fail locally nor when I use the PR docker container on the cluster.

**Comment by jjsjann123** [comment]
I don't have a repro neither. But the issue looks too real to ignore. Looks like some life time issue that causes us to access invalid space. My guess would be to get an ASAN-build might be of help.

**Review by jjsjann123 — CHANGES_REQUESTED** [review]
blocking merge before we clear CI failures.

**Comment by jjsjann123** [comment]
Looks like serde also messes with the python tests. Serde would just use cached kernel rather then having nvfuser compiling new kernels.

i.e. if I run something like `python python_tests/test_python_frontend.py -v -k test_slice_error_checks` twice in a row, the second one would just fail because we are not compiling new kernels.

Is there a flag to turn it off that I missed?

**Comment by jjsjann123** [comment]
I think we are missing some mutex guard in some methods in FusionCache.

i.e. in the `nvfuser::python_frontend::FusionCache::serialization`, you need to lock the mutex there, in case some other threads releases the existing cache. Similarly on Deserialization, you would also need that.

There are likely other places as well. :cry: Luckily the CI caught some failing cases.

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Review by jjsjann123 — APPROVED** [review]
lifting block since CIs are green now.

**Comment by rdspring1** [comment]
!build

File: python_tests/test_python_frontend.py [hunk 8]
```diff
@@ -28,14 +28,9 @@
...
     compute_contiguity,
     compute_tensor_descriptor,
-    serialize as nv_serialize,
 )
 from nvfuser.pytorch_utils import torch_dtype_to_nvfuser_dtype
 
-# Test automatic serialization to common workplace
-import atexit
-
-atexit.register(nv_serialize)
```
[the remark this answers is not in the document]
Comment (rdspring1) [reply python_tests/test_python_frontend.py:38]: I updated this part. We'll clear the `FusionCache` before running the tests in a `setUpModule` function.

**Pull request merged**